src/stream.rs

```diff
@@ -56,129 +37,50 @@ impl<T> From<TlsStream<T>> for MaybeHttpsStream<T> {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        match *self {
-            MaybeHttpsStream::Http(ref mut s) => s.write(buf),
-            MaybeHttpsStream::Https(ref mut s) => s.write(buf),
-        }
-    }
-
-    #[inline]
-    fn flush(&mut self) -> io::Result<()> {
-        match *self {
-            MaybeHttpsStream::Http(ref mut s) => s.flush(),
-            MaybeHttpsStream::Https(ref mut s) => s.flush(),
-        }
-    }
-}
-
-impl<T: AsyncRead + AsyncWrite> AsyncRead for MaybeHttpsStream<T> {
+impl<T: AsyncRead + AsyncWrite + Unpin> AsyncRead for MaybeHttpsStream<T> {
     #[inline]
     unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
-        match *self {
-            MaybeHttpsStream::Http(ref s) => s.prepare_uninitialized_buffer(buf),
-            MaybeHttpsStream::Https(ref s) => s.prepare_uninitialized_buffer(buf),
+        match self {
+            MaybeHttpsStream::Http(s) => s.prepare_uninitialized_buffer(buf),
+            MaybeHttpsStream::Https(s) => s.prepare_uninitialized_buffer(buf),
         }
     }
 
     #[inline]
-    fn read_buf<B: BufMut>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
-        match *self {
-            MaybeHttpsStream::Http(ref mut s) => s.read_buf(buf),
-            MaybeHttpsStream::Https(ref mut s) => s.read_buf(buf),
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        match Pin::get_mut(self) {
+            MaybeHttpsStream::Http(s) => Pin::new(s).poll_read(cx, buf),
+            MaybeHttpsStream::Https(s) => Pin::new(s).poll_read(cx, buf),
         }
     }
 }
 
-impl<T: AsyncWrite + AsyncRead> AsyncWrite for MaybeHttpsStream<T> {
+impl<T: AsyncWrite + AsyncRead + Unpin> AsyncWrite for MaybeHttpsStream<T> {
     #[inline]
-    fn shutdown(&mut self) -> Poll<(), io::Error> {
-        match *self {
-            MaybeHttpsStream::Http(ref mut s) => s.shutdown(),
-            MaybeHttpsStream::Https(ref mut s) => s.shutdown(),
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        match Pin::get_mut(self) {
+            MaybeHttpsStream::Http(s) => Pin::new(s).poll_write(cx, buf),
+            MaybeHttpsStream::Https(s) => Pin::new(s).poll_write(cx, buf),
         }
     }
 
-    #[inline]
-    fn write_buf<B: Buf>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
-        match *self {
-            MaybeHttpsStream::Http(ref mut s) => s.write_buf(buf),
-            MaybeHttpsStream::Https(ref mut s) => s.write_buf(buf),
-        }
-    }
-}
-
-// ===== impl TlsStream =====
-
-impl<T> TlsStream<T> {
-    pub(crate) fn new(inner: native_tls::TlsStream<T>) -> Self {
-        TlsStream {
-            inner,
-        }
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        Poll::Ready(Ok(()))
```

Comment:
Is there a reason that this doesn't defer its implementation to the inner writer like the other methods?

Reply:
Missed by mistake. Fixed now. Thanks!